proposals/custom-rtts/Overview.md

```diff
@@ -376,39 +376,86 @@ C |- br_on_cast_desc_fail l rt_1 rt_2 : t* rt_1 (ref null exact_1 y) -> t* rt_2
 In JS engines,
 WebAssembly RTTs correspond to JS shape descriptors.
 Custom descriptors act as first-class handles to the engine-managed RTTs,
-so they can serve as extension points for the JS reflection of the Wasm objects they describe.
-
-We introduce a new `WebAssembly.setDescriptorPrototype()` API
-that takes a custom descriptor instance and a prototype object.
-The provided prototype will be attached to the descriptor's RTT
-and will become the prototype for the JS reflection
-of all Wasm objects the custom descriptor describes.
-
-The following is a full example that uses `WebAssembly.setDescriptorPrototype()`
+so they can serve as extension points for the JS reflection of the Wasm objects they describe,
+and in particular they can allow JS prototypes to be associated with the described objects.
+To make this work, we allow information about the intended JS reflection of Wasm objects
+to be imported into a Wasm module and held by custom descriptors.
+The `[[GetPrototypeOf]]` algorithm for a WebAssembly object can then look up this information
+on the object's custom descriptor. The details of how this works are described below.
+
+We introduce a new `WebAssembly.DescriptorOptions` type
+that holds relevant information about the JS reflection of Wasm objects.
+A `DescriptorOptions` is constructed with an option bag containing
+an object to be used as a prototype. Other options may be added in the future,
+for example to expose Wasm struct fields as own properties.
+
+```webidl
+dictionary DescriptorOptionsOptions {
+  object? prototype;
+};
+
+[LegacyNamespace=WebAssembly, Exposed=*]
+interface DescriptorOptions {
+  constructor(DescriptorOptionsOptions options);
+}
+```
+
+A `DescriptorOptions` object has a `[[WebAssemblyDescriptorOptions]]`
+internal slot with the value `true`.
+This allows it to be identified by the `[[GetPrototypeOf]]` algorithm.
+Its constructor copies all of the options into the constructed `DescriptorOptions`.
+
+The specification of the `[[GetPrototypeOf]]` internal method
+of an Exported GC Object `O` is updated to perform the following steps
+(which will be made more precise in the final spec):
+
+ 1. If `O.[[ObjectKind]]` is not "struct":
+   1. Return `null`.
```

Comment:
When would this branch be hit?

Reply:
When it is "array" instead.

Reply:
Ah, okay. Seems reasonable to extend this mechanism to provide a prototype for all arrays as well, but seems to good to do in a follow-up, since we'd likely want to reflect arrays of machine integer types as TypedArrays.

Reply:
Yep, we are leaving the door open to extending both the core Wasm part of this and the JS embedding part of this to arrays in the future.